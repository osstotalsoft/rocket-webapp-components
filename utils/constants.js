export const emptyArray = []
export const emptyObject = {}
export const emptyString = ''
export const emptyFunction = () => {}
