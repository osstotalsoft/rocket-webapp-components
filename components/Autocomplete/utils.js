import { createFilterOptions } from '@material-ui/lab/Autocomplete'
import { prop, map, innerJoin, find, propEq, all, includes, is, isEmpty, isNil, props, omit } from 'ramda'
import { emptyArray } from '../../utils/constants'

export const findFirstNotNil = (propNames, option) => find(x => !isNil(x), props(propNames, option))
export const isStringOrNumber = option => is(String, option) || is(Number, option)

const hasStringOptions = options => all(is(String), options) && !isEmpty(options)

const filter = createFilterOptions()
export const filterOptions = (labelKey, valueKey, creatable) => (options, params) => {
  const filtered = filter(options, params)
  const { inputValue } = params

  // Suggest the creation of a new value if it's not empty and it doesn't already exist
  if (creatable && inputValue !== '' && !find(propEq(labelKey, inputValue), options)) {
    filtered.push(
      hasStringOptions(options)
        ? {
            _primitiveValue: inputValue,
            _createdOption: true
          }
        : {
            [valueKey]: inputValue,
            [labelKey]: inputValue,
            _createdOption: true
          }
    )
  }

  return filtered
}

export const getSimpleValue = (options, value, valueKey, isMultiSelection) => {
  if (isMultiSelection && (!is(Array, value) || isEmpty(options))) return emptyArray
  if (!all(is(Object), options)) return value

  // Add new options if the Autocomplete is multiSelection and creatable
  if (is(Array, value)) {
    const optionsSimpleValues = map(prop(valueKey), options)
    value?.map(v => {
      if (!includes(v, optionsSimpleValues)) options.push({ [valueKey]: v })
    })
  }

  const result = isMultiSelection ? innerJoin((o, v) => o[valueKey] === v, options, value) : find(propEq(valueKey, value), options)
  return result || null
}

export const computeChangedMultiValue = (input, simpleValue, valueKey, labelKey) =>
  simpleValue
    ? input.map(a => (is(String, a) ? a : findFirstNotNil([valueKey, labelKey, '_primitiveValue'], a)))
    : input.map(a => (is(String, a) ? a : prop('_primitiveValue', a) || omit(['_createdOption'], a)))

export const computeChangedSingleValue = (input, simpleValue, valueKey, labelKey) =>
  simpleValue ? findFirstNotNil([valueKey, labelKey], input) : prop('_primitiveValue', input) ?? omit(['_createdOption'], input)
