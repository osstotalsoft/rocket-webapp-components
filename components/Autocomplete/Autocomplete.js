import React, { useCallback, useEffect, useState } from "react";
import MuiAutocomplete, {
  createFilterOptions
} from "@material-ui/lab/Autocomplete";
import PropTypes from "prop-types";
import autoCompleteStyles from "./autocompleteStyle";
import { Chip, makeStyles, Checkbox } from "@material-ui/core";
import {
  flatten,
  prop,
  map,
  innerJoin,
  find,
  propEq,
  has,
  all,
  omit,
  contains
} from "ramda";
import { emptyArray, emptyString } from "../../utils/constants";
import CustomTextField from "../CustomTextField";
import Typography from "../Typography";
import cx from "classnames";
import humps from "humps";
import { CheckBoxOutlineBlank, CheckBox } from "@material-ui/icons";

function deprecated(propType, explanation) {
  return function validate(props, propName, componentName, ...rest) {
    if (props[propName] != null) {
      let warned = {};
      const message = `"${propName}" property of "${componentName}" has been deprecated.\n${
        explanation ? explanation : emptyString
      }`;
      if (!warned[message]) {
        console.warn(message);
        warned[message] = true;
      }
    }

    return propType(props, propName, componentName, ...rest);
  };
}

const useStyles = makeStyles(autoCompleteStyles);

const isPrimitive = value =>
  typeof value === "string" ||
  typeof value === "number" ||
  typeof value === "boolean";

const filter = createFilterOptions();
const filterOptions = (labelKey, valueKey, creatable, createdLabel) => (
  options,
  params
) => {
  const filtered = filter(options, params);
  // Suggest the creation of a new value
  if (creatable && params.inputValue !== "") {
    filtered.push(
      isPrimitive(options[0])
        ? {
            primitiveValue: params.inputValue,
            createdLabel: createdLabel
              ? `${createdLabel} '${params.inputValue}'`
              : params.inputValue
          }
        : {
            [valueKey]: undefined,
            [labelKey]: params.inputValue,
            createdLabel: createdLabel
              ? `${createdLabel} '${params.inputValue}'`
              : params.inputValue
          }
    );
  }

  return filtered;
};

const getSimpleValue = (options, value, valueKey, isMultiSelection) => {
  if (isMultiSelection && !Array.isArray(value)) return emptyArray;

  const hasGroups = all(has("options"), options);
  const flattenOptions = hasGroups
    ? flatten(map(prop("options"), options))
    : options;

  // Add new options if the Autocomplete is multiSelection and creatable
  if (Array.isArray(value)) {
    const flattenOptionsSimpleValues = map(prop(valueKey), flattenOptions);
    value?.map(v => {
      if (!contains(v, flattenOptionsSimpleValues))
        flattenOptions.push({ [valueKey]: v });
    });
  }

  const result = isMultiSelection
    ? innerJoin((o, v) => o[valueKey] === v, flattenOptions, value)
    : find(propEq(valueKey, value), flattenOptions);

  return result || null;
};

const getOptionLabel = labelKey => option => {
  if (isPrimitive(option)) return option;
  return option[labelKey] ? option[labelKey] : emptyString;
};

const getOptionSelected = (option, value) => {
  return isPrimitive(option)
    ? option === value
    : JSON.stringify(option) === JSON.stringify(value);
};

const Autocomplete = ({
  options: receivedOptions,
  defaultOptions,
  loadOptions,
  onChange,
  creatable,
  onMenuOpen,
  value,
  isMultiSelection,
  withCheckboxes,
  isSearchable,
  isClearable,
  disabled,
  simpleValue,
  label,
  valueKey,
  labelKey,
  error,
  helperText,
  createdLabel,
  typographyContentColor,
  inputSelectedColor,
  ...other
}) => {
  const classes = useStyles();

  const typographyClasses = cx({
    [classes[`color${humps.pascalize(typographyContentColor)}`]]:
      typographyContentColor !== "initial"
  });

  const [options, setOptions] = useState(
    receivedOptions || (Array.isArray(defaultOptions) && defaultOptions)
  );

  const handleOpen = useCallback(() => {
    if (
      loadOptions &&
      options.length ===
        (Array.isArray(defaultOptions) ? defaultOptions.length : 0)
    )
      loadOptions().then(loadedOptions => {
        setOptions(loadedOptions || emptyArray);
      });
    if (onMenuOpen) onMenuOpen();
  }, [defaultOptions, loadOptions, onMenuOpen, options.length]);

  const renderInput = useCallback(
    params => {
      params.inputProps.className = `${params.inputProps.className} ${classes.input}`;
      if (inputSelectedColor)
        params.inputProps.style = { color: inputSelectedColor };

      const textFieldProps = {
        label,
        error,
        helperText
      };

      return (
        <CustomTextField
          fullWidth
          {...params}
          customInputProps={params.InputProps}
          startAdornment={params.InputProps.startAdornment}
          endAdornment={params.InputProps.endAdornment}
          {...textFieldProps}
        />
      );
    },
    [classes.input, error, helperText, inputSelectedColor, label]
  );

  const renderOption = useCallback(
    (option, { selected }) =>
      withCheckboxes ? (
        <>
          <Checkbox
            icon={<CheckBoxOutlineBlank fontSize="small" />}
            checkedIcon={<CheckBox fontSize="small" />}
            style={{ marginRight: 8 }}
            checked={selected}
          />
          {option.createdLabel ? option.createdLabel : option[labelKey]}
        </>
      ) : (
        <Typography className={classes.input}>
          {isPrimitive(option)
            ? option
            : option.createdLabel
            ? option.createdLabel
            : option[labelKey]}
        </Typography>
      ),
    [classes.input, labelKey, withCheckboxes]
  );

  const renderTags = useCallback(
    (value, getTagProps) =>
      value.map((option, index) => (
        <Chip
          key={index}
          label={
            isPrimitive(option)
              ? option
              : option.primitiveValue
              ? option.primitiveValue
              : option[labelKey]
          }
          {...getTagProps({ index })}
        />
      )),
    [labelKey]
  );

  const handleChange = useCallback(
    (_event, inputValue) => {
      if (!inputValue) return;
      if (isPrimitive(inputValue)) return onChange(inputValue);
      if (isMultiSelection)
        return onChange(
          simpleValue
            ? inputValue.map(a => (a[valueKey] ? a[valueKey] : a[labelKey]))
            : inputValue.map(a =>
                isPrimitive(a)
                  ? a
                  : a.primitiveValue
                  ? a.primitiveValue
                  : omit(["createdLabel"], a)
              )
        );
      if (simpleValue)
        return onChange(
          inputValue[valueKey] ? inputValue[valueKey] : inputValue[labelKey]
        );
      return onChange(
        inputValue.primitiveValue
          ? inputValue.primitiveValue
          : omit(["createdLabel"], inputValue)
      );
    },
    [isMultiSelection, labelKey, onChange, simpleValue, valueKey]
  );

  useEffect(() => {
    if (defaultOptions === true) handleOpen();
  }, [defaultOptions, handleOpen]);

  useEffect(() => {
    setOptions(receivedOptions || emptyArray);
  }, [receivedOptions]);

  return (
    <MuiAutocomplete
      autoComplete
      autoHighlight
      size="small"
      classes={{
        noOptions: `${classes.noOptionsMessage} ${typographyClasses}`
      }}
      forcePopupIcon
      label={label}
      disabled={disabled}
      onOpen={handleOpen}
      options={options || emptyArray}
      handleHomeEndKeys
      selectOnFocus
      clearOnBlur
      disableCloseOnSelect={isMultiSelection}
      filterSelectedOptions={simpleValue && isMultiSelection && !withCheckboxes}
      filterOptions={filterOptions(labelKey, valueKey, creatable, createdLabel)}
      getOptionLabel={getOptionLabel(labelKey)}
      getOptionSelected={!simpleValue ? getOptionSelected : undefined}
      value={
        simpleValue
          ? getSimpleValue(options, value, valueKey, isMultiSelection)
          : value
      }
      multiple={isMultiSelection}
      onChange={handleChange}
      disableClearable={!isClearable}
      renderOption={renderOption}
      renderInput={renderInput}
      renderTags={renderTags}
      freeSolo={creatable}
      {...other}
    />
  );
};

Autocomplete.defaultProps = {
  options: [],
  defaultOptions: [],
  isMultiSelection: false,
  withCheckboxes: false,
  isClearable: false,
  disabled: false,
  simpleValue: false,
  valueKey: "id",
  labelKey: "name",
  error: false,
  value: null,
  creatable: false,
  typographyContentColor: "textSecondary"
};

Autocomplete.propTypes = {
  /**
   * If true, the portion of the selected suggestion that has not been typed by the user,
   * known as the completion string, appears inline after the input cursor in the textbox.
   * The inline completion string is visually highlighted and has a selected state.
   */
  autoComplete: PropTypes.bool,
  /**
   * If true, the first option is automatically highlighted.
   */
  autoHighlight: PropTypes.bool,
  /**
   * The array of options from which the client can select a value.
   */
  options: PropTypes.array,
  /**
   * Function that returns a promise, which is the set of options to be used once the promise resolves.
   */
  loadOptions: PropTypes.func,
  /**
   * The selected value from list of options.
   */
  value: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array,
    PropTypes.number,
    PropTypes.string,
    PropTypes.bool
  ]),
  /**
   * Handle change events on the autocomplete.
   */
  onChange: PropTypes.func.isRequired,
  /**
   * Handle the menu opening.
   */
  onMenuOpen: PropTypes.func,
  /**
   * If true, the user can select multiple values from list.
   */
  isMultiSelection: PropTypes.bool,
  /**
   * If true, the options list will have checkboxes.
   */
  withCheckboxes: PropTypes.bool,
  /**
   * If true, the user can clear the selected value.
   */
  isClearable: PropTypes.bool,
  /**
   * This prop is deprecated, because the new implementation of Autocomplete is searchable by default.
   */
  isSearchable: deprecated(PropTypes.bool),
  /**
   * If true, the Autocomplete is free solo, meaning that the user input is not bound to provided options
   */
  creatable: PropTypes.bool,
  /**
   * If true, the Autocomplete is disabled.
   */
  disabled: PropTypes.bool,
  /**
   * If true, options will be an array of simple values, instead of objects.
   */
  simpleValue: PropTypes.bool,
  /**
   * Label to be displayed in the heading component.
   */
  label: PropTypes.string,
  /**
   * The key of values from options.
   */
  valueKey: PropTypes.string,
  /**
   * The key of the displayed label for each option.
   */
  labelKey: PropTypes.string,
  /**
   * The text to appear below the input, usually coming from a validation.
   */
  helperText: PropTypes.string,
  /**
   * If true, the helper text is displayed when an error pops up.
   */
  error: PropTypes.bool,
  /**
   * The value of label when a new option is added.
   */
  createdLabel: PropTypes.string,
  /**
   * The default set of options to show before the user starts searching. When set to true, the results for loadOptions('') will be autoloaded.
   */
  defaultOptions: PropTypes.oneOfType([PropTypes.bool, PropTypes.array]),
  /**
   * The color of both the text displayed when there are no options and placeholder. It supports those theme colors that make sense for this component.
   */
  typographyContentColor: PropTypes.oneOf([
    "initial",
    "inherit",
    "primary",
    "secondary",
    "textPrimary",
    "textSecondary",
    "error"
  ]),
  /**
   * The color of selected input.
   */
  inputSelectedColor: PropTypes.string
};

export default Autocomplete;
