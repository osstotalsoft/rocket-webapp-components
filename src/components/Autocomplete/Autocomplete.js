import React, { useCallback, useEffect, useState } from "react";
import PropTypes from "prop-types";
import Option from "./Option";
import { LinearProgress, Chip, makeStyles } from "@material-ui/core";
import MuiAutocomplete from "@material-ui/lab/Autocomplete";
import { is, isNil, equals, isEmpty, any, prop } from "ramda";
import {
  filterOptions,
  getSimpleValue,
  findFirstNotNil,
  isStringOrNumber,
  computeChangedMultiValue,
  computeChangedSingleValue
} from "./utils";
import { emptyArray, emptyString } from "../../utils/constants";
import CustomTextField from "../CustomTextField";
import autoCompleteStyle from "./autocompleteStyle";
import cx from "classnames";
import humps from "humps";

const useStyles = makeStyles(autoCompleteStyle);

const Autocomplete = ({
  options: receivedOptions,
  defaultOptions,
  loadOptions,
  loading: receivedLoading,
  loadingText,
  getOptionLabel,
  onChange,
  onInputChange,
  creatable,
  onMenuOpen,
  onClose,
  value,
  isMultiSelection,
  withCheckboxes,
  isClearable,
  disabled,
  simpleValue,
  label,
  valueKey,
  labelKey,
  error,
  helperText,
  required,
  createdLabel,
  typographyContentColor,
  inputSelectedColor,
  isSearchable,
  getOptionDisabled,
  placeholder,
  ...other
}) => {
  const classes = useStyles();

  const typographyClasses = cx({
    [classes[`color${humps.pascalize(typographyContentColor)}`]]:
      typographyContentColor !== "initial"
  });

  const [options, setOptions] = useState(receivedOptions);
  const [asyncOptions, setAsyncOptions] = useState(
    is(Array, defaultOptions) ? defaultOptions : emptyArray
  );

  const [localLoading, setLocalLoading] = useState(false);
  const loading = receivedLoading || localLoading;

  const [localInput, setLocalInput] = useState();
  const [optionsLoaded, setOptionsLoaded] = useState(false);

  const disabledOptions = getOptionDisabled
    ? options.filter(getOptionDisabled)
    : emptyArray;
  const disabledValues = disabledOptions.map(prop(valueKey));
  const isValueDisabled = getOptionDisabled
    ? any(equals(value), disabledValues)
    : false;

  const handleLoadOptions = useCallback(
    input => {
      if (loadOptions) {
        setLocalLoading(true);
        loadOptions(input).then(loadedOptions => {
          setAsyncOptions(loadedOptions || emptyArray);
          setLocalLoading(false);
          setOptionsLoaded(true);
        });
      }
    },
    [loadOptions]
  );

  const handleMenuOpen = useCallback(() => {
    !optionsLoaded && handleLoadOptions(localInput);
    onMenuOpen && onMenuOpen();
  }, [handleLoadOptions, localInput, onMenuOpen, optionsLoaded]);

  const handleMenuClose = useCallback(() => {
    setLocalInput(emptyString);
    onClose && onClose();
  }, [onClose]);

  const renderInput = useCallback(
    params => {
      params.inputProps.className = `${params.inputProps.className} ${classes.input}`;
      if (inputSelectedColor)
        params.inputProps.style = { color: inputSelectedColor };
      params.inputProps.readOnly = !isSearchable;

      const textFieldProps = {
        label,
        error,
        helperText,
        required,
        placeholder
      };

      return (
        <CustomTextField
          fullWidth
          {...params}
          startAdornment={params.InputProps.startAdornment}
          endAdornment={params.InputProps.endAdornment}
          {...textFieldProps}
          InputProps={{ ...params.InputProps, margin: "none" }}
          InputLabelProps={{
            ...params.InputLabelProps,
            margin: null,
            classes: { root: classes.labelRoot, shrink: classes.labelShrink }
          }}
        />
      );
    },
    [
      classes.input,
      classes.labelRoot,
      classes.labelShrink,
      inputSelectedColor,
      isSearchable,
      label,
      error,
      helperText,
      required,
      placeholder
    ]
  );

  const handleOptionLabel = useCallback(
    option => {
      if (getOptionLabel) return getOptionLabel(option);
      if (isStringOrNumber(option)) return option.toString();

      const label = findFirstNotNil([labelKey, valueKey], option);
      return label?.toString() ?? emptyString;
    },
    [getOptionLabel, labelKey, valueKey]
  );

  const renderOption = useCallback(
    (option, { selected }) => {
      const optionLabel = handleOptionLabel(option);

      return (
        <Option
          createdLabel={option._createdOption ? createdLabel : emptyString}
          optionLabel={optionLabel}
          selected={selected}
          withCheckboxes={withCheckboxes}
        />
      );
    },
    [handleOptionLabel, createdLabel, withCheckboxes]
  );

  const renderTags = useCallback(
    (value, getTagProps) =>
      value.map((option, index) => (
        <Chip
          key={index}
          label={handleOptionLabel(option)}
          {...getTagProps({ index })}
          disabled={getOptionDisabled && getOptionDisabled(option)}
        />
      )),
    [getOptionDisabled, handleOptionLabel]
  );

  const getOptionSelected = useCallback(
    (option, value) =>
      simpleValue
        ? option[valueKey] === value[valueKey]
        : equals(option, value),
    [simpleValue, valueKey]
  );

  const handleChange = useCallback(
    (event, inputValue, reason) => {
      // when multi-value and clearable, doesn't clear disabled options that have already been selected
      if (reason === "clear" && getOptionDisabled && isMultiSelection) {
        return onChange(
          computeChangedMultiValue(
            disabledOptions,
            simpleValue,
            valueKey,
            labelKey
          )
        );
      }

      setLocalInput(handleOptionLabel(inputValue));
      // for multi-value Autocomplete, options dialog remains open after selection and we do not want to display a loading state
      loadOptions && !isMultiSelection && setLocalLoading(true);

      if (isNil(inputValue) || isStringOrNumber(inputValue))
        return onChange(inputValue);

      if (isMultiSelection) {
        return onChange(
          computeChangedMultiValue(inputValue, simpleValue, valueKey, labelKey)
        );
      }

      return onChange(
        computeChangedSingleValue(inputValue, simpleValue, valueKey, labelKey)
      );
    },
    [
      disabledOptions,
      getOptionDisabled,
      handleOptionLabel,
      isMultiSelection,
      labelKey,
      loadOptions,
      onChange,
      simpleValue,
      valueKey
    ]
  );

  const handleInputChange = useCallback(
    (event, value) => {
      setLocalInput(value ? value : emptyString);
      onInputChange && onInputChange(event, value);

      // this prevents the component from calling loadOptions again when the user clicks outside it and the menu closes
      if (event?.nativeEvent?.type === "focusout") {
        setOptionsLoaded(false);
        return;
      }

      if (loadOptions) {
        setLocalLoading(true);
        handleLoadOptions(value);
      }
    },
    [handleLoadOptions, loadOptions, onInputChange]
  );

  useEffect(() => {
    // when simpleValue is false, loadOptions has already been called at this point by handleInputChange
    if (!simpleValue || !loadOptions) return;
    if (is(Array, defaultOptions) && !isEmpty(defaultOptions)) return;
    if (defaultOptions === false) return;

    const hasInitialValue = is(Array, value) ? !isEmpty(value) : value;
    // when simpleValue is true, we need to previously load a set of options in order to match the value with one of them
    if (defaultOptions === true || hasInitialValue) {
      handleLoadOptions();
    }
    // this effect should run only at component mount
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    setOptions(receivedOptions || emptyArray);
  }, [receivedOptions]);

  return (
    <MuiAutocomplete
      size="small"
      classes={{
        noOptions: `${classes.noOptionsMessage} ${typographyClasses}`
      }}
      forcePopupIcon
      label={label}
      disabled={disabled || isValueDisabled}
      loading={loading}
      loadingText={loadingText}
      onOpen={handleMenuOpen}
      onClose={handleMenuClose}
      clearOnBlur={!creatable}
      options={
        loading
          ? emptyArray
          : loadOptions
          ? asyncOptions
          : options || emptyArray
      }
      autoHighlight
      handleHomeEndKeys
      selectOnFocus
      disableCloseOnSelect={isMultiSelection}
      filterSelectedOptions={simpleValue && isMultiSelection && !withCheckboxes}
      filterOptions={filterOptions(labelKey, valueKey, creatable, createdLabel)}
      getOptionLabel={handleOptionLabel}
      getOptionSelected={getOptionSelected}
      getOptionDisabled={getOptionDisabled}
      value={
        simpleValue
          ? getSimpleValue(
              loadOptions ? asyncOptions : options,
              value,
              valueKey,
              isMultiSelection
            )
          : value
      }
      multiple={isMultiSelection}
      onChange={handleChange}
      onInputChange={handleInputChange}
      disableClearable={!isClearable}
      renderOption={renderOption}
      renderInput={renderInput}
      renderTags={renderTags}
      freeSolo={creatable}
      {...other}
    />
  );
};

Autocomplete.defaultProps = {
  options: emptyArray,
  defaultOptions: emptyArray,
  isMultiSelection: false,
  isSearchable: true,
  withCheckboxes: false,
  isClearable: false,
  disabled: false,
  simpleValue: false,
  valueKey: "id",
  labelKey: "name",
  error: false,
  required: false,
  value: null,
  creatable: false,
  typographyContentColor: "textSecondary",
  loadingText: <LinearProgress />,
  noOptionsText: "No options"
};

Autocomplete.propTypes = {
  /**
   * The array of options from which the client can select a value.
   */
  options: PropTypes.array,
  /**
   * Function that returns a promise, which resolves to the set of options to be used once the promise resolves.
   */
  loadOptions: PropTypes.func,
  /**
   * If true, the component is in a loading state.
   * By default, this shows a linear progress instead of options.
   * This can be changed by sending the loadingText prop to Autocomplete.
   */
  loading: PropTypes.bool,
  /**
   * Text/component to display when in a loading state.
   */
  loadingText: PropTypes.node,
  /**
   * Text to display when there are no options.
   */
  noOptionsText: PropTypes.node,
  /**
   * Used to determine the string value for a given option.
   */
  getOptionLabel: PropTypes.func,
  /**
   * The selected value from list of options.
   */
  value: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array,
    PropTypes.number,
    PropTypes.string,
    PropTypes.bool
  ]),
  /**
   * Handle change events on the autocomplete.
   */
  onChange: PropTypes.func.isRequired,
  /**
   * Callback fired when the input value changes.
   */
  onInputChange: PropTypes.func,
  /**
   * Handle the menu opening.
   */
  onMenuOpen: PropTypes.func,
  /**
   * Handle the menu closing.
   */
  onClose: PropTypes.func,
  /**
   * If true, the user can select multiple values from list.
   */
  isMultiSelection: PropTypes.bool,
  /**
   * If true, the options list will have checkboxes.
   */
  withCheckboxes: PropTypes.bool,
  /**
   * If true, the user can clear the selected value.
   */
  isClearable: PropTypes.bool,
  /**
   * If false, the user cannot type in Autocomplete, filter options or create new ones.
   */
  isSearchable: PropTypes.bool,
  /**
   * If true, the Autocomplete is free solo, meaning that the user input is not bound to provided options and can add
   * his own values.
   */
  creatable: PropTypes.bool,
  /**
   * If true, the Autocomplete is disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Used to determine the disabled state for a given option.
   */
  getOptionDisabled: PropTypes.func,
  /**
   * If true, options will be an array of simple values, instead of objects.
   */
  simpleValue: PropTypes.bool,
  /**
   * Label to be displayed in the heading component.
   */
  label: PropTypes.string,
  /**
   * The key of values from options.
   */
  valueKey: PropTypes.string,
  /**
   * The key of the displayed label for each option.
   */
  labelKey: PropTypes.string,
  /**
   * The content of the helper under the input.
   */
  helperText: PropTypes.node,
  /**
   * If true, the helper text is displayed when an error pops up.
   */
  error: PropTypes.bool,
  /**
   * Text to be displayed as a placeholder in the text field.
   */
  placeholder: PropTypes.string,
  /**
   * Marks the input field as required (with an *).
   */
  required: PropTypes.bool,
  /**
   * The value of label when a new option is added.
   */
  createdLabel: PropTypes.string,
  /**
   * The default set of options to show before the user starts searching. When set to true, the results for loadOptions('') will be autoloaded.
   */
  defaultOptions: PropTypes.oneOfType([PropTypes.bool, PropTypes.array]),
  /**
   * The color of both the text displayed when there are no options and placeholder. It supports those theme colors that make sense for this component.
   */
  typographyContentColor: PropTypes.oneOf([
    "initial",
    "inherit",
    "primary",
    "secondary",
    "textPrimary",
    "textSecondary",
    "error"
  ]),
  /**
   * The color of selected input.
   */
  inputSelectedColor: PropTypes.string
};

export default Autocomplete;
