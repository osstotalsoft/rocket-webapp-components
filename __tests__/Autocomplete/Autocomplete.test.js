import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import Autocomplete from "../../components/Autocomplete/Autocomplete";
import MuiAutocomplete from "@material-ui/lab/Autocomplete";
import { emptyFunction } from "../../utils/constants";

const options = [
  { id: "1", name: "Approved" },
  { id: "2", name: "Cancelled" }
];

const primitiveOptions = ["Approved", "Cancelled"];

describe("Autocomplete", () => {
  it("snapshot", () => {
    const component = renderer.create(
      <Autocomplete
        fullWidth
        label={"Status"}
        options={options}
        simpleValue
        isClearable
        value={""}
        onChange={jest.fn()}
      />
    );

    const json = component.toJSON();

    expect(json).toMatchSnapshot();
  });

  it("renders all the options", () => {
    const wrapper = mount(
      <Autocomplete
        label={"Status"}
        options={options}
        simpleValue
        value={"2"}
        open
        onChange={jest.fn()}
      />
    );

    expect(wrapper.find("li").length).toEqual(options.length);
  });

  it("the correct option is selected", () => {
    const wrapper = mount(
      <Autocomplete
        label={"Status"}
        options={options}
        simpleValue
        value={"2"}
        open
        onChange={jest.fn()}
      />
    );

    expect(
      wrapper
        .find("li")
        .at(1)
        .props()["aria-selected"]
    ).toEqual(true);
  });

  it("renders the clear button when clearable", () => {
    const wrapper = mount(
      <Autocomplete
        label={"Status"}
        options={options}
        isClearable
        simpleValue
        value={"2"}
        open
        onChange={jest.fn()}
      />
    );

    expect(
      wrapper.find("button.MuiAutocomplete-clearIndicator").length
    ).toEqual(1);
  });

  it("loads options at mount when defaultOptions is true", () => {
    const loadOptions = jest
      .fn()
      .mockImplementation(() => new Promise(emptyFunction));

    mount(
      <Autocomplete
        label={"Status"}
        simpleValue
        onChange={jest.fn()}
        loadOptions={loadOptions}
        defaultOptions
      />
    );

    expect(loadOptions).toBeCalled;
  });

  it("doesn't load options at mount if defaultOptions is not true", () => {
    const loadOptions = jest
      .fn()
      .mockImplementation(() => new Promise(emptyFunction));

    mount(
      <Autocomplete
        label={"Status"}
        simpleValue
        onChange={jest.fn()}
        loadOptions={loadOptions}
      />
    );

    expect(loadOptions.mock.calls.length).toBe(0);
  });

  it("when simpleValue is falsy and the options are objects, displays value[labelKey] in the input", () => {
    const wrapper = mount(
      <Autocomplete
        label={"Status"}
        options={options}
        value={options[0]}
        onChange={jest.fn()}
      />
    );

    const input = wrapper.find("input").at(0);

    expect(input.props().value).toEqual("Approved");
  });

  it("works for primitive options", () => {
    const wrapper = mount(
      <Autocomplete
        label={"Status"}
        options={primitiveOptions}
        value={"Cancelled"}
        open
        onChange={jest.fn()}
      />
    );

    const input = wrapper.find("input").at(0);

    expect(
      wrapper
        .find("li")
        .at(1)
        .props()["aria-selected"]
    ).toEqual(true);
    expect(input.props().value).toEqual("Cancelled");
  });

  it("calls the onChange function provided as a prop", () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Autocomplete
        label={"Status"}
        options={options}
        simpleValue
        isClearable
        value={"2"}
        onChange={onChange}
      />
    );

    const muiAutocomplete = wrapper.find(MuiAutocomplete);
    muiAutocomplete.props().onChange({ target: { value: "1" } });

    expect(onChange).toBeCalled;
  });
});
