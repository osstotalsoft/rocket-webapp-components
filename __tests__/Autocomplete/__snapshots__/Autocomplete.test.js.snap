// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete snapshot 1`] = `
<div
  aria-expanded={false}
  aria-owns={null}
  className="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
  label="Status"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="combobox"
>
  <div
    className="MuiFormControl-root MuiTextField-root undefined undefined MuiFormControl-fullWidth"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root makeStyles-label-21 MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
      data-shrink={false}
      id="undefined-label"
    >
      Status
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd MuiInputBase-marginDense MuiInput-marginDense"
      onClick={[Function]}
    >
      <input
        aria-activedescendant={null}
        aria-autocomplete="both"
        aria-controls={null}
        aria-invalid={false}
        autoCapitalize="none"
        autoComplete="off"
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input makeStyles-input-22 MuiAutocomplete-input MuiAutocomplete-inputFocused makeStyles-input-2 MuiInputBase-inputAdornedEnd MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
        disabled={false}
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        required={false}
        spellCheck="false"
        type="text"
        value=""
      />
      <div
        className="MuiAutocomplete-endAdornment"
      >
        <button
          aria-label="Clear"
          className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Clear"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Open"
          className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Open"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
